{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "BL.SC.U4AIE24224"
      ],
      "metadata": {
        "id": "OE5HlbsLcP2O"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "JAYAVEER VENKATA SAI MADDALI"
      ],
      "metadata": {
        "id": "tnic7rbRdxxJ"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "COMPLEMENT CODE"
      ],
      "metadata": {
        "id": "SNkZzi8LcY0d"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def complement(dna_sequence):\n",
        "    complement = \" \"\n",
        "    for base in dna_sequence:\n",
        "        if base == 'A':\n",
        "            complement += 'T'\n",
        "        elif base == 'T':\n",
        "            complement += 'A'\n",
        "        elif base == 'C':\n",
        "            complement += 'G'\n",
        "        elif base == 'G':\n",
        "            complement += 'C'\n",
        "    return complement\n",
        "\n",
        "dna_sequence = \"ACGTCACGTGT\"\n",
        "print(complement(dna_sequence))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "8Fte6q0pcTQ3",
        "outputId": "1d8523db-b915-490d-dc5a-46f86df27667"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " TGCAGTGCACA\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rxETRpEWQOyz",
        "outputId": "355e1196-0c5f-4234-a9a5-94956cad5828"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n",
            "\u001b[?25l   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/3.2 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[91m━━━━━━━━━━━━\u001b[0m\u001b[90m╺\u001b[0m\u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.0/3.2 MB\u001b[0m \u001b[31m31.5 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m47.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.86\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "USING BIOPYTHON"
      ],
      "metadata": {
        "id": "39r5uug9dMa1"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9WqokGPOQCGI",
        "outputId": "4c7b4bb1-89eb-4539-9c60-96640a4dd33e"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Original DNA Sequence: ACGTCACGTGT\n",
            "Complement Sequence: TGCAGTGCACA\n"
          ]
        }
      ],
      "source": [
        "from Bio.Seq import Seq\n",
        "\n",
        "my_dna=Seq(\"ACGTCACGTGT\")\n",
        "\n",
        "complement = my_dna.complement()\n",
        "\n",
        "print(\"Original DNA Sequence:\", my_dna)\n",
        "print(\"Complement Sequence:\", complement)\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**REVERSE** **COMPLEMENT**"
      ],
      "metadata": {
        "id": "a2E-QHSxQnsd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "dna = \"ACGTCACGTGT\"\n",
        "\n",
        "complement = {\n",
        "    \"A\": \"T\",\n",
        "    \"T\": \"A\",\n",
        "    \"C\": \"G\",\n",
        "    \"G\": \"C\"\n",
        "}\n",
        "\n",
        "dna_complement = \"\"\n",
        "for letter in dna:\n",
        "  dna_complement=dna_complement+complement[letter]\n",
        "\n",
        "print(\"3' -->5' \",dna_complement)\n",
        "\n",
        "reverse_dna_complement = dna_complement[::-1]\n",
        "print(\"5' -->3' \",reverse_dna_complement)\n",
        "rna=reverse_dna_complement.replace(\"T\",\"U\")\n",
        "print(\"5' -->3' \",rna)\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jeC1O_QmQwU_",
        "outputId": "3bb0b82e-a60b-4d84-be8d-05d400c5e7aa"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "TGCAGTGCACA\n",
            "ACACGTGACGT\n",
            "ACACGUGACGU\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "TRANSCRIPTION"
      ],
      "metadata": {
        "id": "zsxEM-BiTNK6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def transcription(dna_sequence):\n",
        "    transcription = \" \"\n",
        "    for base in dna_sequence:\n",
        "        if base == 'T':\n",
        "            transcription += 'U'\n",
        "        elif base == 'A':\n",
        "            transcription += 'A'\n",
        "        elif base == 'C':\n",
        "            transcription += 'C'\n",
        "        elif base == 'G':\n",
        "            transcription += 'G'\n",
        "    return transcription\n",
        "\n",
        "dna_sequence = \"ACGTCACGTGT\"\n",
        "print(transcription(dna_sequence))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7Fj69Y7zR-Co",
        "outputId": "834b1c1d-04c0-4d13-ee9b-b1d36ac38cbc"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            " ACGUCACGUGU\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "USING BIOPYTHON"
      ],
      "metadata": {
        "id": "A9D_pTRBdH2B"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.Seq import Seq\n",
        "\n",
        "my_dna=Seq(\"ACGTCACGTGT\")\n",
        "\n",
        "transcript = my_dna.transcribe()\n",
        "\n",
        "print(\"Original DNA Sequence:\", my_dna)\n",
        "print(\"transcribed rna sequenceSequence:\", transcript)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1VnaQV09VdQq",
        "outputId": "357c5bf2-35f4-486a-f50d-1e143ef2802e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Original DNA Sequence: ACGTCACGTGT\n",
            "transcription Sequence: ACGUCACGUGU\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "CODON TABLE"
      ],
      "metadata": {
        "id": "Si5Yp5jSXJ6g"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def translate_rna_to_protein(rna_sequence):\n",
        "  codon_table={\n",
        "      'UUU': 'F', 'UUC': 'F', 'UUA': 'L', 'UUG': 'L',\n",
        "      'UCU': 'S', 'UCC': 'S', 'UCA': 'S', 'UCG': 'S',\n",
        "      'UAU': 'Y', 'UAC': 'Y', 'UAA': '*', 'UAG': '*',\n",
        "      'UGU': 'C', 'UGC': 'C', 'UGA': '*', 'UGG': 'W',\n",
        "      'CUU': 'L', 'CUC': 'L', 'CUA': 'L', 'CUG': 'L',\n",
        "      'CCU': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
        "      'CAU': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
        "      'CGU': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
        "      'AUU': 'I', 'AUC': 'I', 'AUA': 'I', 'AUG': 'M',\n",
        "      'ACU': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
        "      'AAU': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
        "      'AGU': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
        "      'GUU': 'V', 'GUC': 'V', 'GUA': 'V', 'GUG': 'V',\n",
        "      'GCU': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
        "      'GAU': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
        "      'GGU': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
        "  }\n",
        "\n",
        "  protein_sequence = \"\"\n",
        "  for i in range(0, len(rna_sequence), 3):\n",
        "    codon = rna_sequence[i:i+3]\n",
        "\n",
        "    if len(codon) < 3:\n",
        "      break\n",
        "\n",
        "    amino_acid=codon_table.get(codon)\n",
        "    if amino_acid == '*':\n",
        "      break\n",
        "\n",
        "    if amino_acid is not None:\n",
        "      protein_sequence += amino_acid\n",
        "  return protein_sequence\n",
        "\n",
        "rna_sequence = \"AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUA\"\n",
        "protein_sequence = translate_rna_to_protein(rna_sequence)\n",
        "print(protein_sequence)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qBXU5D3QXSDW",
        "outputId": "2030b6bf-edf9-4f1a-c6ce-01358a853858"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "MAMAPRTEINSTRI\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "USING BIOPYTHON"
      ],
      "metadata": {
        "id": "YrpsAFBKbMsL"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.Seq import Seq\n",
        "\n",
        "my_rna=Seq(\"AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUA\")\n",
        "\n",
        "protein = my_rna.translate()\n",
        "\n",
        "print(\"Original RNA Sequence:\", my_rna)\n",
        "print(\"Protein Sequence:\", protein)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "mKGEzjaZbPff",
        "outputId": "c6f20ed9-8f56-44e7-916a-e6cf350f0a05"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Original RNA Sequence: AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUA\n",
            "Protein Sequence: MAMAPRTEINSTRI\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.12/dist-packages/Bio/Seq.py:2877: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
            "  warnings.warn(\n"
          ]
        }
      ]
    }
  ]
}